### PR TITLE
C64-Secure ABI demo.

### PR DESCRIPTION
**Pull request by jacobbramley**
This demo illustrates a few calling-convention options for weak (but potentially fast) compartmentalisation across function calls.

### CODE REVIEW

File: morello-c64-secure/c64-secure.c [hunk 1]
```diff
@@ -0,0 +1,417 @@
+// Copyright (c) 2024 The CapableVMs "CHERI Examples" Contributors.
+// See COPYRIGHT in the project root for details.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+// This demo is interactive, and should be self-explanatory in use. The code in
+// this file is mostly related to display and infrastructure (such as drawing
+// the backtrace).
+
+#include <assert.h>
+#include <cheri/cheric.h>
```
[the remark this answers is not in the document]
Comment (jacobbramley) [reply morello-c64-secure/c64-secure.c:10]: Oh, good point! No, it's just based on an old prototype and I forgot to port it. Thanks!

**Comment by ltratt** [comment]
@jacobbramley This looks good, and useful -- thanks!

@0152la Is there an easy way of only building this example in CI on Morello? I must admit that I've forgotten how to navigate the infrastructure for this repo...

**Comment by jacobbramley** [comment]
> @0152la Is there an easy way of only building this example in CI on Morello? I must admit that I've forgotten how to navigate the infrastructure for this repo...

I think I could add it to `.buildbot.sh` fairly easily. I'll do that whilst I fix the other issues.

**Comment by ltratt** [comment]
@jacobbramley I remember there's some fiddling because we and try and support multiple platforms. Good luck!

**Comment by 0152la** [comment]
> > @0152la Is there an easy way of only building this example in CI on Morello? I must admit that I've forgotten how to navigate the infrastructure for this repo...
> 
> I think I could add it to `.buildbot.sh` fairly easily. I'll do that whilst I fix the other issues.

I think the process is:
(a) to build, it'll need to be added to `.buildbot.sh`. Currently we seem to build both `riscv` and `morello` together [1], so you'll probably need another block that'll be (pseudocode) `push; make morello-purecap clean; make morello-purecap all; pop`.
(b) to run, add it somewhere in the appropriate block in `run_tests.sh` [2]. I think since it's built only for morello, it'll only be run for morello, but unsure here.

However, since the infrastructure is a bit flimsy I'd think, perhaps as long as it builds it's fine, or even could preclude adding it. I think there's some stuff that's not been ran in CI, and we should probably add everything once (if) we redo the CI.

[1] https://github.com/capablevms/cheri-examples/blob/master/.buildbot.sh#L16-L25
[2] https://github.com/capablevms/cheri-examples/blob/master/tests/run_tests.sh#L64

File: .buildbot.sh [hunk 2]
```diff
@@ -12,7 +12,7 @@ echo "Checking clang-format..."
 CLANG_FORMAT=$CHERI_DIR/morello-sdk/bin/clang-format
 find . -iname "*.c" -o -iname "*.h" -o -iname "*.cpp" -o -iname "*.hpp" | xargs "$CLANG_FORMAT" --dry-run -Werror
 
-echo "Checking that all the purecap examples build for all platforms..."
+echo "Checking that all examples build for all platforms..."
```
Comment (0152la) [inline .buildbot.sh:15]: I'm not sure if removing purecap here is fine, since this block should only be building purecap examples.

File: morello-c64-secure/impls.s [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        .size Z, Z_end-Z
+
+
+        // Helpers for mapping a code address to a given demo function. This is
+        // how we identify which ABI was used.
+
+        .global is_in_A
+        .type is_in_A, %function
+is_in_A:
+        adr c1, A
```
[the remark this answers is not in the document]
Comment (0152la) [reply morello-c64-secure/impls.s]: I think this is done in eed4d19bdd235fc401e5cd85158652f94fd1ba1c?

File: morello-c64-secure/c64-secure.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (0152la) [reply morello-c64-secure/c64-secure.c]: Is this done in d9b2fb6b8ce480373eb0df72a4354bb45a894c44?

File: morello-c64-secure/impls.s (hunk 3 above)
[the remark this answers is not in the document]
Comment (0152la) [reply morello-c64-secure/impls.s]: I think this is done in eed4d19bdd235fc401e5cd85158652f94fd1ba1c?

File: .buildbot.sh [hunk 4]
```diff
@@ -33,6 +33,15 @@ for dir in hybrid hybrid/ddc_compartment_switching syscall-restrict; do
     popd
 done
 
+echo "Checking that all the Morello purecap examples build..."
+platform='morello-purecap'
+for dir in morello-c64-secure; do
```
Comment (probablytom) [inline .buildbot.sh:38]: Perhaps I'm missing something here; why not add `morello-c64-secure` to the list of directories on line 18 (at time of writing), which also runs purecap tests?

I wonder whether it's because this example is Morello-specific and has no `riscv64-purecap` variant to run, in which case maybe [the earlier comment from @0152la](https://github.com/capablevms/cheri-examples/pull/96/files#r1574520120) could be resolved by stating that only riscv64 examples are being built (rather than removing `purecap` altogether in the echo'd text), and building morello targets here by duplicating the list of examples in the earlier `for` loop and removing `morello-purecap` from the list of platforms it builds.

**Review by probablytom — COMMENTED** [review]
Thanks for the patience with this review being so late. Mostly small notes on comments & explanatory text! Great example.

File: morello-c64-secure/c64-secure.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jacobbramley) [reply morello-c64-secure/c64-secure.c]: Yes. Sorry, I should have pointed to that.

File: morello-c64-secure/impls.s (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacobbramley) [reply morello-c64-secure/impls.s]: Yes, and I also added a comment to explain how the tail-calls work.

File: morello-c64-secure/impls.s [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        // TODO: For now, we just allow 16KB per function, but we could pick the
+        // longest length that guarantees 16-byte alignment (E=4).
+        mov x16, #(16 * 1024)
+        sub x17, sp, x16
+        scvalue c17, cfp, x17
+        scbndse c17, c17, x16
+        sub x16, sp, w0, uxtw #4    // Allocate locals.
+        scvalue csp, c17, x16
+        // TODO: The very last section of stack will be unusable because
+        // scbnds{e} fails if the required bounds exceed the bounds of <Cn>.
```
[the remark this answers is not in the document]
Comment (jacobbramley) [reply morello-c64-secure/impls.s:139]: Not before merge, since what is implemented is sufficient for the demo, but we might decide to address them in some future update if the corner-cases are interesting.

I normally use "TODO" for such things, but if the project prefers a different marker (like `NB`), that's fine too.

File: morello-c64-secure/impls.s [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+        // We'll round that down to get the representable length.
+        scoff c10, csp, xzr
+        sub x11, fp, x10
+        // rrlen rounds the length _up_, so use rrmask instead and round the
+        // length down manually.
+        rrmask x12, x11
+        and x12, x11, x12
+        // We assume that since the original stack must have been larger than
+        // this, its base is suitably aligned.
+        // TODO: Prove that this is sound.
```
[the remark this answers is not in the document]
Comment (jacobbramley) [reply morello-c64-secure/impls.s:85]: I don't intend to prove it, no! I'm confident enough (based on intuition) that it's reasonable to build a prototype, but a real proof would be a lot more work. I'll reword this TODO.

File: morello-c64-secure/c64-secure.c [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+		return true;
+	}
+	return false;
+}
+
+// From a given context, ask the user what to call next.
+//
+// This should be called from each demo function in a loop. The demo function
+// should either call the resulting function, or return (exiting its loop) if
+// it's NULL.
```
[the remark this answers is not in the document]
Comment (jacobbramley) [reply morello-c64-secure/c64-secure.c:426]: That's a good suggestion, thanks!

File: .buildbot.sh (hunk 4 above)
Comment (jacobbramley) [reply .buildbot.sh]: I went with your suggestion to run a separate set of tests for each platform: [73485f1](https://github.com/capablevms/cheri-examples/pull/96/commits/73485f1e63609c6be691b323fbb89dfb5b01b056)

Another approach is to iterate through every directory, and build everything that has a makefile. That'd be much simpler, but I worried that an incorrectly-named (or missing) makefile might never be spotted.

File: morello-c64-secure/impls.s (hunk 6 above)
[the remark this answers is not in the document]
Comment (jacobbramley) [reply morello-c64-secure/impls.s]: Reworded: [512da36](https://github.com/capablevms/cheri-examples/pull/96/commits/512da360225968b135d9abdb3986bf77ec7dcbd6)

File: .buildbot.sh (hunk 2 above)
Comment (jacobbramley) [reply .buildbot.sh]: They're cleaned up a bit in [73485f1](https://github.com/capablevms/cheri-examples/pull/96/commits/73485f1e63609c6be691b323fbb89dfb5b01b056).

File: morello-c64-secure/impls.s (hunk 5 above)
[the remark this answers is not in the document]
Comment (probablytom) [reply morello-c64-secure/impls.s:140]: I think `NB` would be clearer! Like you say, what's implemented is fine for the demo itself.

File: morello-c64-secure/impls.s (hunk 6 above)
[the remark this answers is not in the document]
Comment (probablytom) [reply morello-c64-secure/impls.s]: Brill! TY 👍️

File: morello-c64-secure/c64-secure.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (probablytom) [reply morello-c64-secure/c64-secure.c]: This looks brill! I'd nit-pick the `TODO` that snuck in for the reasons mentioned in earlier comments (you suggested `NB` as an alternative in another thread, which I really like) but the reword itself is ace :)

File: .buildbot.sh (hunk 4 above)
Comment (probablytom) [reply .buildbot.sh]: Looks great to me! Cheers! I think some extra indentation might have snuck into the `morello-purecap` bit on two lines, but not exactly sure and it doesn't matter anyway, so marking resolved 👍️

**Comment by ltratt** [comment]
@probablytom Can you mark your comments as resolved (if they're resolved) or comment further? I have the impression this is more-or-less mergeable now, but I'm not 100% sure!

**Comment by jacobbramley** [comment]
Ok, I pushed a change that replaces `TODO` with `NB` in a few places. I think that's all the outstanding comments.

I had already integrated this with `.buildbot.sh` but it's not added to `runtests.sh` because it's purely interactive, and there's no suitable mode for scripted use. It could be scripted, but a failure is likely to manifest as a timeout, so it makes more sense to have a special batch mode, but then that's more code, etc. Let me know if you still think it's worthwhile.

**Comment by ltratt** [comment]
I think we should get this in now, and if we can polish it further in the future, even better!

Please squash.

**Comment by 0152la** [comment]
We don't have CI set up on this repo, so merge queues won't work.
